wheatley/main.py

```diff
@@ -44,46 +45,125 @@ def create_row_generator(args):
 
+    logging.getLogger(Bot.logger_name).setLevel(logging.INFO)
     logging.getLogger(RingingRoomTower.logger_name).setLevel(logging.INFO)
     logging.getLogger(RowGenerator.logger_name).setLevel(logging.INFO)
     logging.getLogger(RegressionRhythm.logger_name).setLevel(logging.INFO)
     logging.getLogger(WaitForUserRhythm.logger_name).setLevel(logging.INFO)
 
 
+def server_main(override_args=None, stop_on_join_tower=False):
+    """
+    The main function of Wheatley when spawned by the Ringing Room server.
+    This has many many fewer options than the standard `main` function, on the basis that this running mode
+    is designed to take its parameters over SocketIO whilst running, rather than from the CLI args.
+    """
+    __version__ = get_version_number()
+
+    parser = argparse.ArgumentParser(
+        description="A bot to fill in bells during ringingroom.com practices (server mode)."
+    )
+
+    parser.add_argument(
+        "room_id",
+        type=int,
+        help="The numerical ID of the tower to join, represented as a row on 9 bells, \
+              e.g. 763451928."
+    )
+    parser.add_argument(
+        "-p", "--port",
+        type=int,
+        help="The port of the SocketIO server (which must be hosted on localhost)."
+    )
+    parser.add_argument(
+        "-l", "--look-to-time",
+        type=float,
+        help="Set to the time when 'Look to' was called if Wheatley was spawned because 'look to' was \
+              called and Wheatley is needed."
+    )
+
+    # Misc arguments
+    parser.add_argument(
+        "--version",
+        action="version",
+        version=f"Wheatley v{__version__}"
+    )
+
+    # Parse arguments
+    # `[1:]` is apparently needed, because sys.argv[0] is the working file of the Python interpreter
+    # which `parser.parse_args` does not want to see as an argument
+    args = parser.parse_args(sys.argv[1:] if override_args is None else override_args)
+
+    # Run the program
+    configure_logging()
+
+    # Args that we are currently 'missing'
+    use_up_down_in = True
+    stop_at_rounds = True
+    peal_speed = 180
+    inertia = 1
+    max_bells_in_dataset = 15
+    handstroke_gap = 1
+    use_wait = True
+
+    try:
+        tower_url = "http://127.0.0.1:" + str(args.port)
+    except TowerNotFoundError:
+        sys.exit("Specified tower not found.")
+
+    tower = RingingRoomTower(args.room_id, tower_url)
+    rhythm = create_rhythm(peal_speed, inertia, max_bells_in_dataset, handstroke_gap, use_wait)
+    bot = Bot(tower, PlaceHolderGenerator(), use_up_down_in, stop_at_rounds, rhythm, user_name="Wheatley",
+              server_mode=True)
+
+    with tower:
+        tower.wait_loaded()
+
+        print("=== LOADED ===")
+
+        if args.look_to_time is not None:
+            bot.look_to_has_been_called(args.look_to_time)
+
+        if not stop_on_join_tower:
+            bot.main_loop()
+
+
 def main(override_args=None, stop_on_join_tower=False):
```

Comment:
Stylistically I have a preference towards creating something like `def console_main(override_args=None, stop_on_join_tower=False):`
so `main(...)` becomes
```
unparsed_args = override_args or sys.argv[1:]
if len(unparsed_args) > 1 and unparsed_args[0] == "server-mode":
    del unparsed_args[0]
    server_main(unparsed_args, stop_on_join_tower)
else:
    console_main(...)
```

Reply:
Oh yes - this is definitely a better plan!  Fixed.